cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+        elif isinstance(indices, slice):
+            start, stop, step = indices.start, indices.stop, indices.step
+            start = start or 0
+            stop = min(stop, self.n) if stop is not None else self.n
+            step = step or 1
+            assert (
+                start >= 0 and stop >= 0 and step >= 0
+            ), "Only positive start/stop/step supported"
+            indices = np.arange(start, stop, step)
+        else:
+            raise TypeError("Unsupported Type for indices")
+
+        assert isinstance(indices, np.ndarray)
+
+        if np.any(indices >= self.n):
+            raise ValueError(f"indices should be < {self.n}")
+
+        # Convert incoming caller indices to indices that this ImageSource will use
+        indices = np.array(self.indices[indices])
+
+        if indices.size == 1:
+            logger.debug(f"ImageSource returning images for index {indices}")
+        else:
+            logger.debug(
+                f"ImageSource returning images for {len(indices)} indices ({indices[0]}..{indices[-1]})"
+            )
+
+        if self.cache:
+            images = self.cache._images(indices)
+        else:
+            images = self._images(indices).astype(self.dtype)
+        return torch.tensor(images)
+
+    def _images(self, indices: np.ndarray) -> np.ndarray:
+        raise NotImplementedError("Subclasses must implement this")
+
+    def chunks(self, chunksize=1000):
+        return ImageSourceIterator(self, chunksize=chunksize)
+
+
+class ArraySource(ImageSource):
+    def __init__(self, data: np.ndarray):
+        if data.ndim == 2:
+            data = data[np.newaxis, ...]
+        nz, ny, nx = data.shape
+        assert ny == nx, "Only square arrays supported"
+        self.data = data
+
+        super().__init__(L=ny, n=nz)
+
+    def _images(self, indices: np.ndarray):
+        return self.data[indices, ...]
+
+
+class MRCFileSource(ImageSource):
+    def __init__(self, filepath: str, *args, **kwargs):
+        from cryodrgn.mrc import MRCHeader
+
+        header = MRCHeader.parse(filepath)
+        self.header = header
+        self.mrcfile_path = filepath
+        self.dtype = header.dtype
+        self.start = 1024 + header.fields["next"]  # start of image data
+        self.nz, self.ny, self.nx = (
+            header.fields["nz"],
+            header.fields["ny"],
+            header.fields["nx"],
+        )
+        assert self.ny == self.nx, "Only square images supported"
+        self.size = self.ny * self.nx
+        self.stride = self.dtype().itemsize * self.size
+
+        super().__init__(L=self.ny, n=self.nz, filenames=filepath, *args, **kwargs)
+
+    def _images(
+        self,
+        indices: np.ndarray,
+        data: Optional[np.ndarray] = None,
+        tgt_indices: Optional[np.ndarray] = None,
+    ):
+
+        with open(self.mrcfile_path) as f:
+            if data is None:
+                data = np.empty((len(indices), self.L, self.L), dtype=self.dtype)
```

Comment:
`empty` is kind of dangerous because it leads to heisenbugs (it will be initialized with 0s most of the time, but sometimes with other values); it can be faster but you've got to be very careful you initialize it. It's safer when non-performance-critical to use something like `np.full` with an invalid value.